### PR TITLE
NMS-14456: Implement Cortex TSS plugin persistance for External Tags

### PR DESCRIPTION
**Pull request by aramosvizcarra**
Added a KeyValueStore to persist external Tags in Postgres and let only Intrinsic and Meta-tags be persited in Cortex.
Key / Values should be Metric key / Metric external Tags
Added an external Cache to cache this Key/Values, similar to the metric Cache.
Added some metrics for Cache usage to see how many Postgres DB transaction can be avoided by Caching.

### CODE REVIEW

File: plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 1]
```diff
@@ -26,15 +27,19 @@
...
 
+    <!--Key-value store -->
+    <reference id="keyValueStore" interface="org.opennms.integration.api.v1.distributed.KeyValueStore" availability="optional"/>
+
     <bean id="timeSeriesStorage" class="org.opennms.timeseries.cortex.CortexTSS" destroy-method="destroy">
         <argument ref="cortexTssConfig" />
+        <argument ref="keyValueStore" />
     </bean>
 
-    <service ref="timeSeriesStorage" interface="org.opennms.timeseries.cortex.CortexTSS" />
```
Comment (aramosvizcarra) [inline plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml:37]: @j-white deleting this line doesn't let the shell command be registered in Karaf. Not sure if this is the reason of that line or there is a proper way to add shell commnds to the registry.

Comment (j-white) [reply plugin/src/main/resources/OSGI-INF/blueprint/blueprint.xml:37]: That makes sense - we should put this one back then

File: README.md [hunk 2]
```diff
@@ -46,6 +46,7 @@ property-set maxConcurrentHttpConnections 100
 property-set writeTimeoutInMs 1000
 property-set readTimeoutInMs 1000
 property-set metricCacheSize 1000
+property-set externalCacheSize 1000
```
Comment (patrick-schweizer) [inline README.md:49]: nit: maybe rename to external**Tag**CacheSize to make it clearer what the cache is about.

File: plugin/src/main/java/org/opennms/timeseries/cortex/CortexTSS.java [hunk 3]
```diff
@@ -213,6 +235,54 @@ public void store(final List<Sample> samples, String clientID) throws StorageExc
...
+            if (externalMetricFromDb.isPresent()) {
+                jsonMetrics = new JSONObject(externalMetricFromDb.get().toString());
+                for (Tag tag : s.getMetric().getExternalTags()) {
+                    if (!jsonMetrics.has(tag.getKey())) {
+                        jsonMetrics.put(tag.getKey(), tag.getValue());
+                        needUpsert = true;
+                    }
+                }
+                if (needUpsert) {
+                    kvStore.putAsync(key, jsonMetrics.toString(), CORTEX_TSS);
```
Comment (patrick-schweizer) [inline plugin/src/main/java/org/opennms/timeseries/cortex/CortexTSS.java:271]: I am not sure but don't we need to also put updated value in cache?
externalTagsCache.put(key, jsonNewMetric.toString());

File: plugin/src/main/java/org/opennms/timeseries/cortex/CortexTSS.java [hunk 4]
```diff
@@ -213,6 +235,54 @@ public void store(final List<Sample> samples, String clientID) throws StorageExc
...
+                    extTagsModified.mark();
+                }
+                //missed caching this record
+                extTagsCacheMissed.mark();
+            } else {
+                for (Tag tag : s.getMetric().getExternalTags()) {
+                    jsonNewMetric.put(tag.getKey(), tag.getValue());
```
Comment (patrick-schweizer) [inline plugin/src/main/java/org/opennms/timeseries/cortex/CortexTSS.java:278]: I am not sure but don't we need to also put updated value in cache?
externalTagsCache.put(key, jsonNewMetric.toString());

Comment (aramosvizcarra) [reply plugin/src/main/java/org/opennms/timeseries/cortex/CortexTSS.java:291]: OK

**Review by patrick-schweizer — APPROVED** [review]
:+1: looks good to me

**Review by cgorantla — APPROVED** [review]
LGTM!

**Pull request merged**